pkg/networkserver/grpc_gsns.go

```diff
@@ -392,7 +375,13 @@ func (ns *NetworkServer) handleUplink(ctx context.Context, up *ttnpb.UplinkMessa
 
 	logger.Debug("Matched device")
 
-	if matched.MACState != nil && matched.MACState.PendingApplicationDownlink != nil {
+	macState := matched.MACState
+	session := matched.Session
+	if pending {
+		macState = matched.PendingMACState
+		session = matched.PendingSession
+	}
```

Comment:
Wouldn't it be more concise to return a struct with device, session and MAC state?

Same reasoning as for the downlink generation

Reply:
That would be a struct like this:
```go
struct {
	Pending bool
	Session *ttnpb.Session
	MACState *ttnpb.MACState
	Device *ttnpb.EndDevice
}
```

Reply:
Yes